src/librustc_mir/transform/validate.rs

```diff
@@ -65,16 +79,107 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                 Rvalue::Use(Operand::Copy(src) | Operand::Move(src)) => {
                     if dest == src {
                         self.fail(
-                            DUMMY_SP,
-                            format!(
-                                "encountered `Assign` statement with overlapping memory at {:?}",
-                                location
-                            ),
+                            location,
+                            "encountered `Assign` statement with overlapping memory",
                         );
                     }
                 }
                 _ => {}
             }
         }
     }
+
+    fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, location: Location) {
+        match &terminator.kind {
+            TerminatorKind::Goto { target } => {
+                self.check_bb(location, *target);
+            }
+            TerminatorKind::SwitchInt { targets, values, .. } => {
+                if targets.len() != values.len() + 1 {
+                    self.fail(
+                        location,
+                        format!(
+                            "encountered `SwitchInt` terminator with {} values, but {} targets (should be values+1)",
+                            values.len(),
+                            targets.len(),
+                        ),
+                    );
+                }
+                for target in targets {
+                    self.check_bb(location, *target);
+                }
+            }
+            TerminatorKind::Drop { target, unwind, .. } => {
+                self.check_bb(location, *target);
+                if let Some(unwind) = unwind {
+                    self.check_bb(location, *unwind);
+                }
+            }
+            TerminatorKind::DropAndReplace { target, unwind, .. } => {
+                self.check_bb(location, *target);
+                if let Some(unwind) = unwind {
+                    self.check_bb(location, *unwind);
+                }
+            }
+            TerminatorKind::Call { func, destination, cleanup, .. } => {
```

Comment:
`Call` (and perhaps others?) could also validate that the destination doesn't overlap with any of the arguments. This would have caught a miscompilation in https://github.com/rust-lang/rust/pull/72632.

Reply:
`DropAndReplace` (and `Yield`?) can do the same check. Not sure about inline assembly, but it seems like that's a candidate for such a check as well.

Reply:
Hm, now I wonder if Miri would have caught that (for `Call`). Unfortunately we can't test Miri on MIR that rustc does not generate.

Reply:
It probably would have, but I haven't yet figured out how to use a locally built miri with external code (like with `cargo-miri`).

Reply:
>  It probably would have, but I haven't yet figured out how to use a locally built miri with external code (like with cargo-miri).

I usually do `./miri install` and then use `cargo miri`.

Reply:
`./src/tools/miri/miri install` fails with "can't find crate for `rustc_apfloat`" for me, and running the miri tests ICEs the compiler

Reply:
Ah I've only ever used the script in a standalone clone... yeah it probably doesn't work in the rustc folder layout.

> running the miri tests ICEs the compiler

How do you run them? `./x.py test --stage 0 src/tools/miri` should work -- except when the Miri toolstate is broken for the commit you're basing your work on.

Reply:
Yeah that's what I tried to run, but it enters unreachable code while decoding or encoding something. Not really sure what's happening, but I'll do more digging.

Reply:
That sounds like an incremental problem or `--keep-stage` gone wrong... strange.